contracts/ProcedureTable.sol

```diff
@@ -411,30 +717,65 @@ library ProcedureTable {
+            uint256 pPointer = _getProcedurePointerByKey(uint192(keys[i]));
+            r[n] = uint192(keys[i]); n++;
             // Store the keyIndex at this location
-            r[n] = p.keyIndex; n++;
-            r[n] = uint256(p.location); n++;
-            // number of capabilities in array
-            r[n] = p.caps.length; n++;
-            // cycle through each capability
-            for (uint248 j = 0; j < p.caps.length; j++) {
-                // we add 1 to account for the capType
-                uint256 capSize = p.caps[j].values.length + 1;
-                r[n] = capSize; n++;
-                r[n] = p.caps[j].capType; n++;
-                for (uint248 k = 0; k < p.caps[j].values.length; k++) {
-                    r[n] = p.caps[j].values[k]; n++;
+            r[n] = _get(pPointer); n++;
+            r[n] = _get(pPointer + 1); n++;
+            // Save this spot to record the the total number of caps
+            uint256 nCapTypesLocation = n; n++;
+            uint256 totalCaps = 0;
+            // Cycle through all cap types [0,255], even though most don't exist
+            for (uint256 j = 1; j <= 10; j++) {
+                // How many caps are there of this type
+                uint256 nCaps = _get(pPointer | (j*0x10000) | 0x00 | 0x00);
+                // Only record the caps if they're at least 1W
+                if (nCaps > 0) {
+                    uint256 capSize = capTypeToSize(j);
+                    // Cycle through them and add them to the array. Here we need to
+                    // know the size.
+                    for (uint256 k = 1; k <= nCaps; k++) {
+                    // record the size
+                    r[n] = (capSize+2); n++;
+                    // record the type
+                    r[n] = j; n++;
+                        totalCaps++;
+                        for (uint256 l = 0; l < capSize; l++) {
+                            r[n] = _get(pPointer | (j*0x10000) | (k*0x100) | (l*0x1)); n++;
+                        }
+                    }
                 }
             }
+            r[nCapTypesLocation] = totalCaps;
         }
+        r[0] = n;
         return r;
     }
 
+    function capTypeToSize(uint256 capType) internal pure returns (uint256) {
+        if (capType == CAP_PROC_CALL) {
+            return 1;
+        } else if (capType == CAP_STORE_WRITE) {
+            return 2;
+        } else if (capType == CAP_LOG) {
+            return 5;
+        } else if (capType == CAP_PROC_REGISTER) {
+            return 1;
+        } else if (capType == CAP_PROC_DELETE) {
+            return 1;
+        } else if (capType == CAP_PROC_ENTRY) {
+            return 0;
+        } else if (capType == CAP_ACC_CALL) {
+            return 1;
+        } else {
+            revert("invalid capability type");
+        }
+
+    }
+
     function insert(Self storage /* self */, bytes24 key, address value, uint256[] caps) internal returns (bool replaced) {
```

Comment:
This uses `#_getProcedureByKey` which is quite expensive.

Reply:
My aim is to remove the Procedure struct completely. If you're ok with that I can remove getProcedureByKey completely, as well as parseCaps.

Reply:
I've just recalled where I went with this. I was aiming to remove those two functions, but then realised they were used by BeakerContract.sol, so I held back. They, are however, no longer necessary for the kernel. I made a new commit that removes these last few references.

I think we should move these functions into BeakerContract.sol, as they are of little use to the kernel.

Reply:
We're using them for tests though, so it should be fine to keep it for now. Just not part of `ProcedureTable.sol`, we could move it to `IKernel.sol` or `KernelTest.sol`

Reply:
Yeah, that was kind of my thinking too. I did just go back in for five minutes and made sure all references to that struct were removed (it didn't actually depend on it anymore). So it still exists, just for the purposes of other places.